### PR TITLE
Add GDScript exports for pre-loaded const Resource scripts and Resource-derived script classes.

### LINKED ISSUE #22641: Add support for exporting resource script constants and script classes.
**Godot version:**
3.1 Alpha 1

**Issue description:**
I would expect the following to work:

    # my_resource.gd
    extends Resource
    class_name MyResource
    
    # node.gd
    extends Node
    export(MyResource) var my_res # ERROR: Expected a constant expression

I can actually confirm it isn't a constant by creating a derived version of `MyResource`, class naming it as `SuperResource`, and then literally re-assigning it, which I really don't think I should be able to do:

    extends Node
    func _ready():
        MyResource = SuperResource
        print(MyResource.resource_path) # prints "res://super_resource.gd"

To ME, this seems like a bug that shouldn't be allowed, and which consequently would allow the export of the type name to work in 3.1.

It also seems like this just flat-out wasn't ever a feature, even in 3.0.

    const MyResource = preload("res://my_resource.gd")
    export(MyResource) var my_res # Error: Export hint not a resource type (even though it is!)

**Steps to reproduce:**
Recreate the code examples above for Resource-extending scripts.

3.0: create a constant and try to export it. Fails.
3.1: Same as 3.0 + create a class name and try to export it. Fails.

### PR DESCRIPTION
**Pull request by willnationsdev**
Fixes #22641, #6763.

Allows you to export resource scripts if they are preloaded constants or script classes. Custom types can't be supported this way since custom type information is not accessible at runtime.

There is a known "bug" if you do either of these though. The EditorInspector won't display the right icon or show a class name in the editor property, so there's something wonky going on there. That can be a follow-up fix though.

### CODE REVIEW

**Comment by willnationsdev** [comment]
Took out the icon method changes in `editor_node.cpp` since they pertain to the separate bug, and I'd rather keep this PR as cherry-pickable as possible.

**Comment by nhydock** [comment]
Instead of using preloaded consts, can strings also be supplied if class_name is used to register the resources elsewhere?  Doing something like 

```
export(MyResource) var template
```
or even
```
export(Resource, "res://my_resource.gd") var template
```

I feel could be equally useful/clean.

**Comment by willnationsdev** [comment]
@nhydock 

> 

> Instead of using preloaded consts, can strings also be supplied if class_name is used to register the resources elsewhere? Doing something like
> 
> ```
> export(MyResource) var template
> ```

See OP.

> Allows you to export resource scripts if they are preloaded constants **or script classes.**

The feature you described is already handled by this PR.

**Comment by willnationsdev** [comment]
@MysteryGM FYI, if you haven't already seen this PR, it's a pretty quick change that comes with some minor UI bugs in 3.1, but either way, you could easily build this into the version of Godot you guys use and it would greatly simplify your guys' workflow.

**Review by bojidar-bg — COMMENTED** [review]
LGTM, but I think it might be better if the hint string contains the path to the script or the global class name. Potentially, the inspector might have to be changed to support that as well, but it will be for the better in any case.

**Comment by samgreen** [comment]
This would be pretty useful for my workflow and seems like it's missing at the moment. 

Just curious, does this also pre-populate an instance of script and allow you to edit from the inspector?

**Comment by willnationsdev** [comment]
@samgreen yes, it does. It is an extremely powerful use-case, so I'm looking forward to eventually getting this (or something like it) merged in.

**Comment by xDGameStudios** [comment]
this could really be cherry picked... hope it does!
Takes away a lot of inspector headache!

btw @willnationsdev does this work for Arrays as well?!

**Comment by LeonardMeagher2** [comment]
Is there any reason this shouldn't be merged? Are there bigger plans related to this?

File: modules/gdscript/gdscript_parser.cpp [hunk 1]
```diff
@@ -5506,6 +5528,12 @@ GDScriptParser::DataType GDScriptParser::_type_from_variant(const Variant &p_val
 		}
 		if (scr.is_valid()) {
 			result.script_type = scr;
+
+			String name = scr->get_language()->get_global_class_name(scr->get_path());
+			if (ScriptServer::is_global_class(name)) {
+				result.is_constant = true;
```
Comment (vnen) [inline modules/gdscript/gdscript_parser.cpp]: I don't understand the need of this. It's already set to `true` before.

**Comment by vnen** [comment]
I'm okay with this (apart from my comment). It needs a rebase to fix conflicts though.

**Comment by willnationsdev** [comment]
Been a while since I've looked at this. Will have to rebase and retest. As I recall, there was a resulting GUI error in the Inspector when I did this, even though it would "work" functionally.

**Comment by nhydock** [comment]
@willnationsdev Should this PR be marked as closed/abandoned in favor of #26162

**Comment by willnationsdev** [comment]
@nhydock yes, I think so.

**Comment by akien-mga** [comment]
Closing as superseded by #26162 then.

**Pull request closed without merging**